<a href="https://colab.research.google.com/github/solangeMaricel/Proyecto3-Digitalers/blob/main/Copy_of_Notebook_7.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

## Programa Ingenias+ Data Science

## Trabajando con Pandas

1) Importa las librerias requeridas: `Pandas` y `NumPy`.  

Una buena practica a la hora de escribir codigo es de comentar siempre que estoy haciendo (usando `#`) en la linea anterior. Por mas obvio que parezca lo que voy a hacer, puede ser que tenga que compartir el documento con otras personas y no sepan que quise hacer en cada linea. O que yo mismo lea el codigo tiempo despues y no recuerde porque lo hice.

In [ ]:
import pandas as pd
import numpy as np

2) Lee el archivo csv `life_expectancy_data` usando `pandas`. Guardalo en una variable llamada `expvida`.

In [ ]:
# Leo los datos en un DataFrame
path = "/content/life_expectancy_data.csv"
expvida = pd.read_csv(path)

3) ¿Que tipo de estructura de datos contiene la variable `expvida`?

In [ ]:
# Obtengo el tipo de estructura de datos
print(expvida.info())
print(type(expvida))

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2938 entries, 0 to 2937
Data columns (total 22 columns):
 #   Column                           Non-Null Count  Dtype  
---  ------                           --------------  -----  
 0   Country                          2938 non-null   object 
 1   Year                             2938 non-null   int64  
 2   Status                           2938 non-null   object 
 3   Life expectancy                  2928 non-null   float64
 4   Adult Mortality                  2928 non-null   float64
 5   infant deaths                    2938 non-null   int64  
 6   Alcohol                          2744 non-null   float64
 7   percentage expenditure           2938 non-null   float64
 8   Hepatitis B                      2385 non-null   float64
 9   Measles                          2938 non-null   int64  
 10   BMI                             2904 non-null   float64
 11  under-five deaths                2938 non-null   int64  
 12  Polio               

4) ¿Cuantas filas y columnas tiene `expvida`?

In [ ]:
# Uso .shape para obtener cantidad de filas y columnas
print(expvida.shape)
print(f"Número de Filas: {expvida.shape[0]}")
print(f"Número de Columnas: {expvida.shape[1]}")

(2938, 22)
Número de Filas: 2938
Número de Columnas: 22


**Numero de filas**: 2938


**Numero de columnas**: 22

5) ¿Cual es el nombre de las columnas contenidas en `expvida`?

In [ ]:
# Obtengo el nombre de las columnas
print(expvida.columns)

Index(['Country', 'Year', 'Status', 'Life expectancy ', 'Adult Mortality',
       'infant deaths', 'Alcohol', 'percentage expenditure', 'Hepatitis B',
       'Measles ', ' BMI ', 'under-five deaths ', 'Polio', 'Total expenditure',
       'Diphtheria ', ' HIV/AIDS', 'GDP', 'Population',
       ' thinness  1-19 years', ' thinness 5-9 years',
       'Income composition of resources', 'Schooling'],
      dtype='object')


6) Inspecciona las primeras 10 filas de `expvida`

In [ ]:
# Inspecciono las primeras filas
print(expvida.head())

       Country  Year      Status  life_expectancy  Adult Mortality  \
0  Afghanistan  2015  Developing             65.0            263.0   
1  Afghanistan  2014  Developing             59.9            271.0   
2  Afghanistan  2013  Developing             59.9            268.0   
3  Afghanistan  2012  Developing             59.5            272.0   
4  Afghanistan  2011  Developing             59.2            275.0   

   infant deaths  Alcohol  percentage expenditure  Hepatitis B  measles  ...  \
0             62     0.01               71.279624         65.0     1154  ...   
1             64     0.01               73.523582         62.0      492  ...   
2             66     0.01               73.219243         64.0      430  ...   
3             69     0.01               78.184215         67.0     2787  ...   
4             71     0.01                7.097109         68.0     3013  ...   

   Polio  Total expenditure  Diphtheria    HIV/AIDS         GDP  Population  \
0    6.0           

7) Y ahora las 10 ultimas.

In [ ]:
# Inspecciono las ultimas filas
print(expvida.tail(10))

8) ¿Que tipos de datos contiene cada una de las columnas de `expvida`?

In [ ]:
# Tipo de datos en cada columna
print(expvida.dtypes)

Country                             object
Year                                 int64
Status                              object
life_expectancy                    float64
Adult Mortality                    float64
infant deaths                        int64
Alcohol                            float64
percentage expenditure             float64
Hepatitis B                        float64
measles                              int64
BMI                                float64
under-five deaths                    int64
Polio                              float64
Total expenditure                  float64
Diphtheria                         float64
 HIV/AIDS                          float64
GDP                                float64
Population                         float64
 thinness  1-19 years              float64
 thinness 5-9 years                float64
Income composition of resources    float64
Schooling                          float64
dtype: object


9) ¿Hay valores faltantes en alguna de las columnas de `expvida`?

In [ ]:
# Obtengo la cantidad de datos faltantes en cada columna
missing_data_count = expvida.isnull().sum()
print(missing_data_count)

Country                              0
Year                                 0
Status                               0
life_expectancy                     10
Adult Mortality                     10
infant deaths                        0
Alcohol                            194
percentage expenditure               0
Hepatitis B                        553
measles                              0
BMI                                 34
under-five deaths                    0
Polio                               19
Total expenditure                  226
Diphtheria                          19
 HIV/AIDS                            0
GDP                                448
Population                         652
 thinness  1-19 years               34
 thinness 5-9 years                 34
Income composition of resources    167
Schooling                          163
dtype: int64


10) Selecciona el maximo de datos faltantes. Calcula que porcentaje de los datos representan los datos faltantes.

In [ ]:
# Obtengo el maximo valor de datos faltantes y lo guardo en la variable masnan
masnan = expvida.isnull().sum().max()
masnan

652

In [ ]:
# Obtengo la cantidad de filas total y divido el maximo valor de datos faltantes por este numero
# y lo multiplico por 100
total_datos = expvida.shape[0] * expvida.shape[1]
porcentaje_faltantes = (masnan / total_datos) * 100

64636
2938


In [ ]:
# Imprimo el maximo porcentaje de datos faltantes. Uso f-strings.
print(f"El máximo de datos faltantes es: {masnan}")
print(f"El porcentaje de datos faltantes es (sobre el total de datos del DataFrame): {round(porcentaje_faltantes, 2)}%")
print(f"El porcentaje de datos faltantes es (sobre el total de datos de la columna): {round(masnan/expvida.shape[0]*100, 2)}%")

El máximo de datos faltantes es: 652
El porcentaje de datos faltantes es (sobre el total de datos del DataFrame): 1.01%
El porcentaje de datos faltantes es (sobre el total de datos de la columna): 22.19%


- str.format()
- Template Class
- f-strings

#### Ahora vamos a realizar algunos calculos y manipulaciones sobre las columnas.

Una función útil para este fin es: `.rename()`. Y podemos utilizarla de la siguiente manera: Si `df` es mi DataFrame y quiero renombrar la columna `columna` como `columna1`, y la columna `columna2` como `columna3`, utilizo la siguiente sintaxis:
  
`df.rename(columns={"columna": "columna1", "columna2": "columna3"}, inplace=True)`

11) Renombra la columna 'Life expectancy ' como 'life_expectancy', ' BMI ' como 'bmi' y ''Measles ' como 'measles'. No olvides de usar `inplace` para modificar el DataFrame original. Luego chequea que el cambio se haya hecho correctamente.

In [ ]:
# Renombro columnas del DataFrame
expvida.rename(columns={'Life expectancy ': "life_expectancy", " BMI ": "BMI", "Measles ": "measles"}, inplace=True)

In [ ]:
# Chequeo que se hayan reemplazado correctamente
print(expvida.columns)

Index(['Country', 'Year', 'Status', 'life_expectancy', 'Adult Mortality',
       'infant deaths', 'Alcohol', 'percentage expenditure', 'Hepatitis B',
       'measles', 'BMI', 'under-five deaths ', 'Polio', 'Total expenditure',
       'Diphtheria ', ' HIV/AIDS', 'GDP', 'Population',
       ' thinness  1-19 years', ' thinness 5-9 years',
       'Income composition of resources', 'Schooling'],
      dtype='object')


12) ¿Cual es el promedio de expectativa de vida (Life expectancy) en general?

In [ ]:
expvida.describe()

Year  life_expectancy  Adult Mortality  infant deaths  \
count  2938.000000      2928.000000      2928.000000    2938.000000   
mean   2007.518720        69.224932       164.796448      30.303948   
std       4.613841         9.523867       124.292079     117.926501   
min    2000.000000        36.300000         1.000000       0.000000   
25%    2004.000000        63.100000        74.000000       0.000000   
50%    2008.000000        72.100000       144.000000       3.000000   
75%    2012.000000        75.700000       228.000000      22.000000   
max    2015.000000        89.000000       723.000000    1800.000000   

           Alcohol  percentage expenditure  Hepatitis B        measles  \
count  2744.000000             2938.000000  2385.000000    2938.000000   
mean      4.602861              738.251295    80.940461    2419.592240   
std       4.052413             1987.914858    25.070016   11467.272489   
min       0.010000                0.000000     1.000000       0.000000   
25%       0.877500                4.685343    77.000000       0.000000   
50%       3.755000               64.912906    92.000000      17.000000   
75%       7.702500              441.534144    97.000000     360.250000   
max      17.870000            19479.911610    99.000000  212183.000000   

               BMI  under-five deaths         Polio  Total expenditure  \
count  2904.000000         2938.000000  2919.000000         2712.00000   
mean     38.321247           42.035739    82.550188            5.93819   
std      20.044034          160.445548    23.428046            2.49832   
min       1.000000            0.000000     3.000000            0.37000   
25%      19.300000            0.000000    78.000000            4.26000   
50%      43.500000            4.000000    93.000000            5.75500   
75%      56.200000           28.000000    97.000000            7.49250   
max      87.300000         2500.000000    99.000000           17.60000   

       Diphtheria      HIV/AIDS            GDP    Population  \
count  2919.000000  2938.000000    2490.000000  2.286000e+03   
mean     82.324084     1.742103    7483.158469  1.275338e+07   
std      23.716912     5.077785   14270.169342  6.101210e+07   
min       2.000000     0.100000       1.681350  3.400000e+01   
25%      78.000000     0.100000     463.935626  1.957932e+05   
50%      93.000000     0.100000    1766.947595  1.386542e+06   
75%      97.000000     0.800000    5910.806335  7.420359e+06   
max      99.000000    50.600000  119172.741800  1.293859e+09   

        thinness  1-19 years   thinness 5-9 years  \
count            2904.000000          2904.000000   
mean                4.839704             4.870317   
std                 4.420195             4.508882   
min                 0.100000             0.100000   
25%                 1.600000             1.500000   
50%                 3.300000             3.300000   
75%                 7.200000             7.200000   
max                27.700000            28.600000   

       Income composition of resources    Schooling  
count                      2771.000000  2775.000000  
mean                          0.627551    11.992793  
std                           0.210904     3.358920  
min                           0.000000     0.000000  
25%                           0.493000    10.100000  
50%                           0.677000    12.300000  
75%                           0.779000    14.300000  
max                           0.948000    20.700000

In [ ]:
# Obtengo el promedio de la columna
median_life_exp = round(expvida['life_expectancy'].mean(),2)
print(f"Promedio de expectativa de vida: {median_life_exp} años")

Promedio de expectativa de vida: 69.22 años


13) ¿Cual es el promedio de expectativa de vida (Life expectancy) por país?

In [ ]:
# Obtengo el promedio por pais
promedio_por_pais = expvida.groupby('Country')['life_expectancy'].mean()
print(promedio_por_pais)


Country
Afghanistan                           58.19375
Albania                               75.15625
Algeria                               73.61875
Angola                                49.01875
Antigua and Barbuda                   75.05625
                                        ...   
Venezuela (Bolivarian Republic of)    73.38750
Viet Nam                              74.77500
Yemen                                 63.86250
Zambia                                53.90625
Zimbabwe                              50.48750
Name: life_expectancy, Length: 193, dtype: float64


14) ¿Cual es la expectativa de vida mas alta y mas baja? Usa la funcion de pandas y la funcion de NumPy y compara el resultado. ¿Que paises tienen esas expectativas de vida?

In [ ]:
# Maxima expectativa de vida usando numpy
max_life_exp_np = np.max(expvida['life_expectancy'])


In [ ]:
# Maxima expectativa de vida usando pandas
max_life_exp = expvida['life_expectancy'].max()


In [ ]:
# Minima expectativa de vida usando numpy
min_exp_life_np = np.min(expvida['life_expectancy'])


36.3

In [ ]:
# Minima expectativa de vida usando pandas
min_life_exp = expvida['life_expectancy'].min()


36.3

In [ ]:
# Muestro los resultados maximos
print("Máxima expectativa de vida usando Pandas:",max_life_exp)
print("Máxima expectativa de vida usando Numpy:",max_life_exp_np)

Máxima expectativa de vida usando Pandas: 89.0
Máxima expectativa de vida usando Numpy: 89.0


In [ ]:
# Muestro los resultados minimos
print("Minima expectativa de vida usando Pandas:",min_life_exp)
print("Minima expectativa de vida usando Numpy:",min_exp_life_np)

Minima expectativa de vida usando Pandas: 36.3
Minima expectativa de vida usando Numpy: 36.3


In [ ]:
# Filtro las columnas para mostrar solo las filas que tengan valor
# igual al maximo
df_filtered_max = expvida[expvida['life_expectancy'] == max_life_exp]
df_filtered_max

Country  Year      Status  life_expectancy  Adult Mortality  \
241       Belgium  2014   Developed             89.0             76.0   
915       Finland  2014  Developing             89.0             78.0   
937        France  2008  Developing             89.0             88.0   
938        France  2007  Developing             89.0             89.0   
995       Germany  2014   Developed             89.0             69.0   
1293        Italy  2004   Developed             89.0             66.0   
1850  New Zealand  2010   Developed             89.0             71.0   
1916       Norway  2009   Developed             89.0             67.0   
2056     Portugal  2014   Developed             89.0             78.0   
2433        Spain  2007   Developed             89.0             72.0   
2513       Sweden  2007   Developed             89.0             63.0   

      infant deaths  Alcohol  percentage expenditure  Hepatitis B  measles  \
241               0    12.60             7163.348923         98.0       70   
915               0     8.80             6164.455402          NaN        0   
937               3    11.90             7002.785925         47.0      604   
938               3    12.20               64.737149         42.0       39   
995               2    11.03              941.756291         88.0      443   
1293              2     8.98             4270.914882         96.0      599   
1850              0     9.62             6664.148708          9.0       43   
1916              0     6.68              142.374943          NaN        2   
2056              0     9.88              271.254553         98.0        0   
2433              2    11.05              510.932701         96.0      267   
2513              0     6.90             7593.391622          NaN        1   

      ...  Polio  Total expenditure  Diphtheria    HIV/AIDS          GDP  \
241   ...   99.0               1.59         99.0        0.1  47439.39684   
915   ...   98.0               9.68         98.0        0.1  49914.61864   
937   ...   98.0               1.57         98.0        0.1  45413.65710   
938   ...   99.0               1.45         98.0        0.1    416.58397   
995   ...   94.0              11.30         95.0        0.1   4792.65288   
1293  ...   97.0               8.49         94.0        0.1  31174.56118   
1850  ...   93.0              11.20         93.0        0.1  33691.34837   
1916  ...   94.0               9.49         94.0        0.1    817.77681   
2056  ...   98.0               9.50         98.0        0.1   2277.53613   
2433  ...   96.0               8.36         96.0        0.1   3279.41400   
2513  ...   98.0               8.92         98.0        0.1  53324.37937   

      Population   thinness  1-19 years   thinness 5-9 years  \
241     112957.0                    1.0                  1.0   
915    5461512.0                    0.9                  0.8   
937    6437499.0                    0.6                  0.6   
938    6416229.0                    0.6                  0.6   
995      89825.0                    1.1                  1.1   
1293  57685327.0                    0.5                  0.5   
1850         NaN                    0.3                  0.3   
1916   4828726.0                    0.7                  0.7   
2056     14162.0                    0.7                  0.5   
2433   4522683.0                    0.6                  0.5   
2513    914892.0                    1.3                  1.3   

      Income composition of resources  Schooling  
241                             0.890       16.3  
915                             0.890       17.0  
937                             0.877       16.1  
938                             0.873       16.1  
995                             0.920       17.0  
1293                            0.846       15.6  
1850                            0.899       20.3  
1916                            0.936       17.4  
2056                            0.837       16.8  
2433

In [ ]:
# Filtro las columnas para mostrar solo las filas que tengan valor
# igual al minimo
df_filtered_min = expvida[expvida['life_expectancy'] == min_life_exp]
df_filtered_min

Country  Year      Status  life_expectancy  Adult Mortality  \
1127   Haiti  2010  Developing             36.3            682.0   

      infant deaths  Alcohol  percentage expenditure  Hepatitis B  measles  \
1127             23     5.76               36.292918          NaN        0   

      ...  Polio  Total expenditure  Diphtheria    HIV/AIDS         GDP  \
1127  ...   66.0                8.9         66.0        1.9  662.279518   

      Population   thinness  1-19 years   thinness 5-9 years  \
1127   9999617.0                    4.0                  4.0   

      Income composition of resources  Schooling  
1127                             0.47        8.6  

[1 rows x 22 columns]

15) ¿Cuantas categorias encontramos en la columna "Status"? _**Pista**_: Lee acerca de la función [unique()](http://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.Series.unique.html?highlight=unique#pandas.Series.unique)

In [ ]:
# Obtengo los valores unicos de la columna status
expvida['Status'].unique()

array(['Developing', 'Developed'], dtype=object)

16) ¿Cuantos paises se ubican en cada una de esas categorias de "Status"? _**Pista**_: Lee acerca de la función [value_counts()](http://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.Series.value_counts.html)

In [ ]:
# Cuento cuantos valores hay de cada categoria
expvida['Status'].value_counts()

Status
Developing    2426
Developed      512
Name: count, dtype: int64

18) ¿Cuantos paises presentaron algun caso de Sarampion (**Measles**) en el 2015? **_Pista_**: Lee acerca de la función [.count()](http://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.count.html?highlight=count#pandas.DataFrame.count)

In [ ]:
# Cuento cuantas filas tienen valores distintos de cero
expvida.count()

Country                            2938
Year                               2938
Status                             2938
life_expectancy                    2928
Adult Mortality                    2928
infant deaths                      2938
Alcohol                            2744
percentage expenditure             2938
Hepatitis B                        2385
measles                            2938
BMI                                2904
under-five deaths                  2938
Polio                              2919
Total expenditure                  2712
Diphtheria                         2919
 HIV/AIDS                          2938
GDP                                2490
Population                         2286
 thinness  1-19 years              2904
 thinness 5-9 years                2904
Income composition of resources    2771
Schooling                          2775
dtype: int64